sdk/program/src/vote/state/vote_state_versions.rs

```diff
@@ -1,8 +1,9 @@
-use super::{vote_state_0_23_5::VoteState0_23_5, *};
+use super::{vote_state_0_23_5::VoteState0_23_5, vote_state_1_10_40::VoteState1_10_40, *};
 
 #[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
 pub enum VoteStateVersions {
     V0_23_5(Box<VoteState0_23_5>),
+    V1_10_40(Box<VoteState1_10_40>),
```

Comment:
I think this will still break snapshots because nodes with this upgrade will 

1. Convert to the new version of vote states here: https://github.com/solana-labs/solana/blob/fb2d4aecc2e4d19f58e5c6cab1b6464913937960/programs/vote/src/vote_state/mod.rs#L996
2. Then set that new state here: https://github.com/solana-labs/solana/blob/fb2d4aecc2e4d19f58e5c6cab1b6464913937960/programs/vote/src/vote_state/mod.rs#L1005

This new v3 state will then show up in snapshots, and validators without the upgrade will break

Reply:
Saw your comment here: https://github.com/solana-labs/solana/pull/28126#discussion_r1032994383. I would rather get the pain of doing this correctly rather than the more easily shippable version

Reply:
I think that snapshots should be versioned; then validators can only search for and download snapshots supported by their software version (or older versions).  Then when an incompatible change to the snapshot format is introduced, the version can be incremented, and only software that can understand that version will load those snapshots (or keep searching for a supported snapshot version from other validators if not).  On incompatible snapshot version change, newer software will be able to load any snapshot, but older software will be increasingly unable to find snapshots until it upgrades version.  That will be even more incentive to upgrade software versions (although downloading snapshots should be rare anyway; validators should be saving their own and loading them).

If this mechanism doesn't exist, then yeah I guess I'll have to code up the snapshot writing to produce the v2 format until the feature is enabled.

Reply:
Anyway, given that VoteState is versioned, this would be an issue for any new version of VoteState.  How is this supposed to be handled?  I looked at the change that introduced the VoteStateVersions concept (d821fd29d6e66aeeb4581ea548caf907c54d9ee0) and I didn't see anything related to retaining snapshot compatibility there, although I could have missed it.

As far as I can tell, the code is not well set up to handle versioning of snapshots, especially versioning that would change at runtime according to a feature flag.

So how should I proceed?

Reply:
Oh wait looking at SnapshotConfig, it does have a "snapshot_version" field.  I guess I need to increment that?  Not sure how that value gets communicated to Vote State serialization during snapshotting though.

Reply:
Or maybe don't need to communicate anything.  Just add a new snapshot version that is a part of the change, so that any software running with the timely vote credits changes will create snapshots with the new version number.

Reply:
I've updated the snapshot version.  I presume that's sufficient.  It will cause older software to refuse to use snapshots from newer software, which will prevent older software from getting incompatible vote accounts.

Is this sufficient?

Reply:
Tag @brooksprumo on the best way to handle this

Reply:
@bji we could guard against storing this new version of the `VoteState` until the feature is enabled. I.e. have the conversion https://github.com/solana-labs/solana/blob/fb2d4aecc2e4d19f58e5c6cab1b6464913937960/programs/vote/src/vote_state/mod.rs#L996 and subsequent store https://github.com/solana-labs/solana/blob/fb2d4aecc2e4d19f58e5c6cab1b6464913937960/programs/vote/src/vote_state/mod.rs#L1005 use the old version until the feature is enabled

Reply:
Tag @brooksprumo double checking the best way to handle this

Reply:
I don't think it's really possible to avoid incompatible snapshot versions.  Even if the new version of vote state is only written into snapshots after the feature is enabled, it will still introduce incompatible versions because software that is old will be able to download snapshots from software that is new and post-feature-enablement, and thus producing incompatible snapshots.

As a validator operator I personally do not see the snapshot incompatibility as a problem, especially if it's done only after feature enablement.  All validators will have ample time to upgrade by the time the feature is enabled.

Reply:
I would be fine with backporting changes to allow older software releases to be able to read snapshots of the newer release.  This would I think be pretty easily accomplished by taking the entire change, but with a different feature id, that is not enabled on that branch.

Reply:
In terms of adding the fields at the end of the snapshot, I don't see how that changes things much.  It's still an incompatible snapshot, it's just incompatible in a different way (extra junk at the end that will still break snapshot ingestion by unaware software versions), and additionally it would I believe result in a lot more complicated snapshot loading code (although I haven't looked at examples of how this was done in the past; but having to fetch account state from disparate pieces of the snapshot sounds complicated).

I think it's better to have the snapshot loading code be simpler (i.e. let vote accounts deserialize their state using the VoteStateVersions mechanism that was already created), and manage the backporting and sequencing of feature enablement (and also implementing the "don't write new vote account state version until the feature is enabled" concept -- which I hope is not too complicated; I haven't looked into exactly what is involved in getting the feature set available to the vote account serialization code, if it's not already there).

Reply:
> @bji we could guard against storing this new version of the `VoteState` until the feature is enabled. I.e. have the conversion
> 
> https://github.com/solana-labs/solana/blob/fb2d4aecc2e4d19f58e5c6cab1b6464913937960/programs/vote/src/vote_state/mod.rs#L996
> and subsequent store
> 
> https://github.com/solana-labs/solana/blob/fb2d4aecc2e4d19f58e5c6cab1b6464913937960/programs/vote/src/vote_state/mod.rs#L1005
> use the old version until the feature is enabled

These are the points at which VoteState is loaded, not saved into a snapshot.  Unfortunately, the VoteState must be upgraded on load, because all of the code is written expecting to operate against the new version.

What must occur is that the VoteState version that is written into snapshots needs to always be the older version until the feature is enabled.  This is OK because until the feature is enabled, the latencies that would be stored in the vote accounts are always 0, and the default which is used when the VoteAccount state is loaded is also 0.

However, using the feature set on Vote Account state serialization into the account is likely to be tricky.  I do not believe that the feature set is easily available at that point and a lot of plumbing will probably be needed to get it there.

Reply:
from the last time i looked at this the local tower is serialized independent of snapshot / ledger, just in a file by itself

Reply:
> I think it's worthwhile to do the snapshot compatibility testing that I mentioned earlier:
> 
>     * Run a validator with this branch and load from a snapshot created by a v1.13 node. Verify the node starts up. (Or at least reconstructs the Bank successfully)
> 
>     * Run a v1.13 node and load from a snapshot created by this branch. Verify the node starts up. (Or at least reconstructs the Bank successfully)
> 
> 
> Do make sure to _remove_ the new `V1_3_0` SnapshotVersion too.
> 
> I was looking through the snapshot itself and didn't see a place where we serialize the tower or votes as either part of the Bank or the AccountsDb, so the snapshot itself may be OK. I see the tower gets serialized to a file, like `tower-1_9-xxx.bin`, but I don't see that file in snapshot archives either. (And on my validator that I run for development/testing I don't ever see a tower bin in my ledger directory.) All that to say that the _snapshot_ may be fine with these changes. And if that's the case, may be solvable with feature gates, which I feel are simpler/more straight-forward.

Yes in fact the snapshots are technically compatible between software versions before and after the change, because there is no change in the snapshot format.  But the vote account state changes between the software versions and this would result in software that doesn't know about the new vote account state successfully loading snapshots from software that does know and has enabled the new vote account state, and immediately failing to work because the vote program instruction executions would all fail.

Anyway, I have most recently made the following changes:

1. Removed the new snapshot version number because I realize that it's not necessary - the snapshot format has not changed
2. Cause the vote program to write only the "old state" version into its accounts until the feature flag is enabled; then it will write the "new state" version.

This will allow snapshots of account state to remain compatible between software versions up until the moment that the feature flag is enabled; but by that point all software will have been upgraded to support the new version and there will be no problems with snapshots.

Reply:
But yeah the code to do it is just ugly, no doubt, I wish it didn't look like a bunch of arbitrary stuff sprinkled throughout.  I promise that once the code is integrated, and the feature enabled, and everyone's been on the same software for a while, I will make a PR to remove all the garbage and clean it up again.